### DISCUSSION TITLE
Scoping state/`CapsuleReader` prop drilling?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#50] [Q&A] [answered]
I ended with some edge scenarios. So in this thread I want to discuss two main things:
- A way to crash if using and old `use`:
I know passing down a `use` object is a bad practice, but have you implemented a mechanism to prevent calling an old `use` object? This would help to prevent leaks/bugs.

- `CapsuleReader`/`CapsuleHandle` prop drilling:
I will try to explain a sceneario where I only see `prop drilling` as a solution.
I have an asbtract class `CreateDialog` that extends of `RearchConsumer`. It allows to build different create dialogs for different models with different fields and field types. It has an abstract `createItem` method. In most of the cases, the specific implementation would need to access a `CapsuleReader` to perform the operation. As in `RearchConsumer` the `WidgetHandle` object is only available within the `build` method, the only way to make it work would be to prop drilling into `createItem` method. Do you agree?
I think this case is not the worst bad practice on prop drilling, but I envision other worst scenarios that I would comment later on other thread.

**Comment by GregoryConrad** [comment, marked as answer]
> A way to crash if using and old use:
I know passing down a use object is a bad practice, but have you implemented a mechanism to prevent calling an old use object? This would help to prevent leaks/bugs.

Hmm, right now there is an `assert` that checks whether or not `use` is invoked outside of a capsule's current build for both reading capsules and side effects, but I didn't ever add a check for whether it is an "old" copy of a use. [Here](https://github.com/GregoryConrad/rearch-dart/blob/445a051d60bc42ddeccef7029d2a43314eeebf16/packages/rearch/lib/src/impl.dart#L93) is what is currently done. That should cover every case that didn't purposefully try to get around that assert (and you would have to *really* try to avoid the assert).

I frankly think that should be enough of a check (someone would have to do something extremely weird to have an old `use` in a new build), but I guess I would be open to PRs for an "old" assertion as well, if the implementation isn't too bad.

This is one of those times I like Rust, because this situation can never even happen in the first place. 

> CapsuleReader/CapsuleHandle prop drilling

See https://rearch.gsconrad.com/flutter/scoping-state

Eventually, there will be a macro like detailed above to avoid all of the excessive boilerplate for a `RearchConsumer` and `InheritedWidget` combo, but for now, you need to manually define a new RearchConsumer and InheritedWidget to handle scoping state/data. (Make a new InheritedWidget that takes its data from its args, and then supply those args via the RearchConsumer which returns the new InheritedWidget.)
